### PR TITLE
gh-123660: Internal macros for accessing empty string/bytes singletons

### PR DESCRIPTION
**Pull request by picnixz**
This is a proposal for using macros instead of typing everytime `_Py_STR(empty)` or `_Py_SINGLETON(bytes_empty)`

This is a follow-up to https://github.com/capi-workgroup/decisions/issues/36#issuecomment-2271622565.

I wanted to use `Py_GetConstantBorrowed` instead, but I wasn't sure which approach was the most efficient one. In `Py_GetConstantBorrowed` you have a `constant_id < Py_ARRAY_LENGTH(constants)` check and I'm not sure it's even possible to use the function since you'll need to create the constant in the array `constants`, and this uses `PyUnicode_New(0, 0)` which itself returns the singleton. So you'll end up with a recursion issue.

Anyway, it's still WIP although all occurrences to empty strings and bytes were replaced. I'm just opening a draft PR so it's easier to see the diff (I'll create an issue later).

* Issue: gh-123660

### CODE REVIEW

**Comment by picnixz** [comment]
@vstinner @encukou @banjoerlend In our discussion on the `PyBytes_Join` function, we discussed on the possibility of using macros for empty strings/bytes. Do you think this is worth the change? if so, I'll open an associated issue. I haven't changed the `PyUnicode_FromString("")` calls since they are "optimized" to returnt he singleton (though there are quite a lot of functions being called under the hood).

**Comment by encukou** [comment]
The C API WG deals with *public* API.

IMO the internal API is fine as it is; the rename doesn't bring much.

**Comment by zooba** [comment]
I'm quite happy to review internal C API designs as well - _we_ all still have to live with them.

As a concept, this one looks fine to me. It'll help our contributors discover how best to access these singletons, as it isn't exactly obvious. If/when we make another way to get them it'll be easier to update.

As a PR, I think there are some generated files you've touched that will be undone when they regenerate (though I haven't checked for sure). Make sure you do a `make regen` to see if anything changes.

**Comment by picnixz** [comment]
> As a PR, I think there are some generated files you've touched that will be undone when they regenerate (though I haven't checked for sure).

I think there's already a CI job that checks whether the files are up-to-date or not and I locally did regenerate them so they should be part of the PR. It *should* be fine I think.

I'll create an issue soon (I think we can make a meta-issue to integrate the empty tuple case as well).

**Comment by picnixz** [comment]
For now, I'm converting it into a draft until I've resolved my own interrogations on whether to replace `PyUnicode_FromString("")` and `PyUnicode_New(0, 0)` or not.

**Comment by vstinner** [comment]
> For now, I'm converting it into a draft until I've resolved my own interrogations on whether to replace PyUnicode_FromString("") and PyUnicode_New(0, 0) or not.

_Py_STR() and _Py_SINGLETON() are replaced with direct pointers at build time, whereas PyUnicode_FromString("") and PyUnicode_New(0, 0) are implemented as regular function calls at runtime: it's slower.

File: Include/internal/pycore_global_strings.h [hunk 1]
```diff
@@ -797,6 +797,11 @@ struct _Py_global_strings {
      ? (PyObject*)&_Py_SINGLETON(strings).ascii[(CH)] \
      : (PyObject*)&_Py_SINGLETON(strings).latin1[(CH) - 128])
 
+#define _Py_DECLARE_EMPTY_STRING    _Py_DECLARE_STR(empty, "")
```
Comment (vstinner) [inline Include/internal/pycore_global_strings.h]: I don't think that _Py_DECLARE_EMPTY_STRING makes sense. Keeping _Py_DECLARE_STR() in unicode_get_empty() should be enough.

If you replace _Py_STR(empty) with _Py_EMPTY_STRING, you should remove _Py_DECLARE_STR(empty, ...) in all places except in unicode_get_empty().

**Comment by picnixz** [comment]
> _Py_STR() and _Py_SINGLETON() are replaced with direct pointers at build time, whereas PyUnicode_FromString("") and PyUnicode_New(0, 0) are implemented as regular function calls at runtime: it's slower.

Yes, but is it safe to replace all occurrences? Actually, what I'd like to know whether there was a deeper meaning of not using the singleton (is it because when the code was written, this singleton did not yet exist? I think it was introduced in 2022).

File: Include/internal/pycore_global_strings.h (hunk 1 above)
Comment (picnixz) [reply Include/internal/pycore_global_strings.h]: Sure! I'll make a separate commit for this in a moment.

Comment (picnixz) [reply Include/internal/pycore_global_strings.h]: Actually `compile.c` also uses the macro in a function. The comment for `_Py_DECLARE_STR` says that it should be used before any invokation of `_Py_EMPTY_STRING` in a *function*. So, technically, we should also keep it (I'll also check that `_Py_EMPTY_STRING` is correctly paired for detection actually).

**Comment by vstinner** [comment]
> Yes, but is it safe to replace all occurrences?

I don't think that it's useful to replace all occurrences.

**Comment by zooba** [comment]
> I don't think that it's useful to replace all occurrences.

Are there some you know are not safe? Provided the reference counting works out (which may require additional changes), it should be fine to replace calls like `PyUnicode_FromString("")` with the empty str singleton, right?

**Comment by encukou** [comment]
> The C API WG deals with public API.

I was extremely brief there, sorry. In the WG discussion, [Marc-Andre asks for public API](https://github.com/capi-workgroup/decisions/issues/36#issuecomment-2263864855).

And I think that is what we need first. Perhaps something like:
- `#define Py_NONE Py_GetConstantBorrowed(Py_CONSTANT_NONE)`
- `#define Py_FALSE Py_GetConstantBorrowed(Py_CONSTANT_FALSE)`
- ... and so on

These should be documented. They're ”borrowed from the interpreter”, i.e they don't necessarily survive interpreter shutdown/reinitialization, and can't be shared between interpreters, although both will work in CPython.
Unlike `Py_None`, `Py_NONE` would be in the limited API, but it would *not* be an exported ABI symbol.

And when that's in place, these can get a faster implementation for non-limited API or for `Py_BUILD_CORE`.

I don't think it's helpful to only add new undocumented internal macros.

**Comment by vstinner** [comment]
I didn't propose `#define Py_EMPTY_STR Py_GetConstantBorrowed(Py_CONSTANT_EMPTY_STR)` since `Py_GetConstantBorrowed()` is documented as:

> This function is **primarily intended for backwards compatibility**: using Py_GetConstant() is recommended for new code.

It's strange to me to add a new macro relying on something only provided "for backward compatibility".

If we add such macro, I would prefer to see the function documentation rephrased.

Or maybe it's a bad idea to add a new macro which returns a borrowed reference, whereas the C API Guidelines is against adding new functions returning borrowed references. I don't know.

**Comment by vstinner** [comment]
@encukou:
> I don't think it's helpful to only add new undocumented internal macros.

I have mixed feelings about this change. I'm not convinced that it makes Python easier to maintain or easier to understand. I like using explicitly the complex _Py_STR() and _Py_SINGLETON() macros, but maybe it's just a personal taste :-)

**Comment by encukou** [comment]
> If we add such macro, I would prefer to see the function documentation rephrased.

I agree with that.

> C API Guidelines is against adding new functions returning borrowed references

Which guidelines? I guess we should clarify those first :)

**Comment by zooba** [comment]
> I like using explicitly the complex _Py_STR() and _Py_SINGLETON() macros, but maybe it's just a personal taste :-)

I suspect this is personal taste, because you are highly familiar with how it works under the covers :-) For someone who isn't so aware, I think they're unlikely to guess to write the more complex version, and so if we have an obvious macros that does it then they're more likely to choose something efficient.

**Comment by picnixz** [comment]
I'm closing this one since we didn't manage to find an answer to what we should replace at the end and whether this macro would be useful or not. We can revisit this question later (and there are anyway lots of conflicts now and I'm cleaning up my stale branches).

**Pull request closed without merging**